test/src/main/java/org/springframework/security/test/web/reactive/server/JwtMutator.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright 2002-2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.security.test.web.reactive.server;
+
+import static org.springframework.security.test.web.reactive.server.SecurityMockServerConfigurers.mockAuthentication;
+
+import org.springframework.http.client.reactive.ClientHttpConnector;
+import org.springframework.lang.Nullable;
+import org.springframework.security.test.support.JwtAuthenticationBuilder;
+import org.springframework.test.web.reactive.server.MockServerConfigurer;
+import org.springframework.test.web.reactive.server.WebTestClient;
+import org.springframework.test.web.reactive.server.WebTestClientConfigurer;
+import org.springframework.web.server.adapter.WebHttpHandlerBuilder;
+
+/**
+ * @author Jérôme Wacongne &lt;ch4mp&#64;c4-soft.com&gt;
+ * @since 5.2
+ */
+public class JwtMutator extends JwtAuthenticationBuilder<JwtMutator>
```

Comment:
We should try and favor composition over inheritance. One way to achieve this would be to, in the `SecurityMockServerConfigurers` and `SecurityMockMvcRequestPostProcessors` classes, doing:

```java
public JwtMutator mockJwt(Consumer<JwtAuthenticationBuilder> jwtAuthenticationBuilderConsumer) {
    JwtAuthenticationBuilder jwtAuthenticationBuilder = new JwtAuthenticationBuilder();
    // ... any defaults necessary
    jwtAuthenticationBuilderConsumer.accept(jwtAuthenticationBuilder);
    return new JwtMutator(jwtAuthenticationBuilder);
}
```

Reply:
Could you please detail a bit more the `JwtMutator` implementation you're thinking of?

> We should try and favor composition over inheritance

I agree with this. But "favor" isn't "avoid", right?

To me `JwtMutator` and `JwtRequestPostProcessor` **are** builders in the sens the "flow" API they implement is builder API: exposed methods incrementally mutate properties later used to build something.

Additionally, I don't quite know how to turn this "are builders" into "have builders" without proxying each and every builder method in both `JwtMutator` and `JwtRequestPostProcessor`. So if you could kindly give me a hint...

Reply:
You are spot on that doing a delegate here would be a pain due to all the replicated methods.

If you expose the builder in `mockJwt()`, though, then you won't need to do that:

```java
this.client.mutateWith(
    mockJwt(jwtAuthenticationBuilder -> {
        jwtAuthenticationBuilder.name("user"); // and whatever else
    }))
    // ...
```

`JwtMutator` (and `JwtRequestPostProcessor` by association) would then look like:

```
public class JwtMutator implements WebTestClientConfigurer, MockServerConfigurer {
    private final JwtAuthenticationBuilder jwtAuthenticationBuilder;

    // constructor, overrides

    private <T extends WebTestClientConfigurer & MockServerConfigurer> T configurer(...) {
        return mockAuthentication(this.jwtAuthenticationBuilder.build());
    }
}
```

Reply:
And actually, I think a potential improvement on that would be to separate out the notion of creating a `Jwt` from creating an authentication token:

```java
this.client.mutateWith(
    mockJwt(jwtBuilder -> jwtBuilder.claim(ISS, "https://idp.example.com"))
            .authorities(...))
    // ...
```

And then:

```java
public class JwtMutator implements WebTestClientConfigurer, MockServerConfigurer {
    private final JwtBuilder jwtBuilder;
    private Collection<GrantedAuthority> authorities = AuthorityUtils.
            .createAuthorityList("SCOPE_USER");

    // constructor, overrides, authorities support

    private <T extends WebTestClientConfigurer & MockServerConfigurer> T configurer(...) {
        return mockAuthentication(new JwtAuthenticationToken(this.jwtBuilder.build(), this.authorities));
    }
}
```

And finally:

```java
public static JwtMutator mockJwt(Consumer<JwtBuilder> jwtBuilderConsumer) {
    JwtBuilder jwtBuilder = new JwtBuilder();
    jwtBuilder.header("alg", "none");
    // ... other defaults
    jwtBuilderConsumer.apply(jwtBuilder);
    return new JwtMutator(jwtBuilder);
}
```

This has the benefit of focusing on the biggest complexity, creating the `Jwt`, and `JwtBuilder` seems really beneficial to promote up to `oauth2-jose`, for example as an inner class to `Jwt`. It would also make `JwtRequestPostProcessor` and `JwtMutator` quite reminiscent of `UserRequestPostProcessor` and `UserExchangeMutator`, which is often a good thing. It does have a very slight consequence of duplicating the authorities support in `JwtRequestPostProcessor` and `JwtMutator`.

Reply:
The more I'm thinking of it, the more I believe that what should be removed is `authorities(...)` from `JwtAuthenticationTokenBuilder` public API:
* authorities are not enough for JWT security: expressions can use claims that won't be converted to authorities (see sample below)
* claims always contain authorities (deserializing it correctly is just a matter of using the app authorities converter)
* storing authorities in addition to claims in `JwtAuthenticationTokenBuilder` leads to the same potential incoherence issues, we have inside `Jwt` for issue and expiration instants: it is bad idea to store two different representations of the same value (authorities). 

**Sample incoherent `JwtAuthenticationTokenBuilder` with claims and authorities**
`builder.claim("scp", "read").authorities(new SimpleGrantedAuthority("SCOPE_write")` will be interpreted differently by 
* `@PreAuthorize("hasAuthority('SCOPE_read'))` with JwtGrantedAuthoritiesConverter
* `@PreAuthorize("#scp.contains('read')")` with `@AuthenticationPrincipal(expression="claims['scp']") String scopes`) 

both expressions would behave the same with a real token

**security expression not involving authorities**
``` java
@RestController
public class OAuth2ResourceServerController {
    @PreAuthorize("#age >= 16")
    @GetMapping("/controlled")
    public String controlled(@AuthenticationPrincipal(expression="claims['age']") Integer age) {
        return "Hey mon, buy me drink.";
    }
}
```
``` java
@RunWith(SpringRunner.class)
@WebMvcTest(OAuth2ResourceServerController.class)
public class OAuth2ResourceServerControllerTest {
	@Autowired
	MockMvc mockMvc;

	@MockBean
	JwtDecoder jwtDecoder;
	
	@Test
	public void controlChecksIsOver16() throws Exception {
		mockMvc.perform(get("/controlled").with(jwt()))
				.andExpect(status().isForbidden());
		
		mockMvc.perform(get("/controlled").with(jwt().claim("age", 18)))
				.andExpect(content().string(is("Hey mon, buy me drink.")));
		
		mockMvc.perform(get("/controlled").with(jwt().claim("age", 15)))
				.andExpect(status().isForbidden());
	}
}
```

Reply:
> The more I'm thinking of it, the more I believe that what should be removed is authorities(...) from JwtAuthenticationTokenBuilder public API:

I think this would be quite surprising since that is one of the constructor parameters for `JwtAuthenticationToken`. It would also be strange since `GrantedAuthority` is the first-class citizen. But, I do believe you are on to something.

I believe instead, doing a `JwtBuilder` makes more sense - it doesn't have `authorities` and it's `claims`-focused. The builder also has the potential to address the coherence concerns you have raised about `Jwt`.

If we do that (see https://github.com/jzheaux/spring-security/commit/44192fdb4890ae686d7a0a7ffe70ab98219516d8 for a rough visual), then `JwtRequestPostProcessor` could potentially also do something like this:

```java
public JwtRequestPostProcessor authorities(
    Converter<Jwt, Collection<GrantedAuthority>> jwtAuthoritiesConverter) {
    this.authorities = jwtAuthoritiesConverter.convert(this.jwt);
}
```

I think it's utility is questionable, and I'd say leave it out of this PR, but I'm open to suggestions.

What I like about it is the tester gets some additional flexibility, while the API still keeps the focus on the first-class domain object, `GrantedAuthority`. It's arguable that such a configuration point isn't necessary in `UserRequestPostProcessor` because `User` itself has a list of authorities, while `Jwt` doesn't.

> builder.claim("scp", "read").authorities(new SimpleGrantedAuthority("SCOPE_write"))

This would certainly be odd for the application to do. A more likely scenario is:

> builder.authorities(new SimpleGrantedAuthority("SCOPE_read"), new SimpleGrantedAuthority("SCOPE_write"))

As, between the two, GrantedAuthority is the first-class citizen.

> @PreAuthorize("hasAuthority('SCOPE_read')")

This is sensible since `hasAuthority` means "has a granted authority whose authority string is `SCOPE_read`"
 
> @PreAuthorize("#scopes.contains('read')")

This would be custom on the part of the application. No `scopes` variable is published by Spring Security. Given that the developer has customized the evaluation context this heavily, I believe we can trust them to do what works for them.

> @AuthenticationPrincipal(expression="claims['scp']") String scopes)

This is certainly possible, though I'd wonder what they are trying to accomplish first by doing something OAuth-specific and second by accessing a primitive that is already translated for them. It would be more sensible to do:

```java
@AuthenticationPrincipal(expression="authorities") Collection<GrantedAuthority> authorities
```

I'm certainly not saying that user's won't find benefit in looking at the "scp" claim, but this notion that I'm going to mix and match in my application is messy, and I don't see why the framework should try and simplify that. Either they'll look up the scope claim or they'll use granted authorities with granted authorities being heavily preferred since it isn't OAuth-specific.

`JwtAuthenticationConverter` (and the associated granted authorities converter) isn't about claim-authority coherence. It's about the fact that an `Authentication` needs a `Collection` of `GrantedAuthority`s so that it can be reasoned about in the application. Providing simple ways to provide those authorities is great, but trying to sync up individual claims with those authorities is not the goal.

Reply:
To me, in OAuth2 world, the tokens are the authorization holders. Spring GrantedAuthority are derived from it on the resource-server.

As so, if in an OAuth2 authentication builder we accept a token:
* we must derive authorities from it
* if we are not able to figure out token authorization from spring GrantedAuthority, then we should not accept it as builder parameter

`JwtAuthenticationToken` constructor accepts a token and a collection of granted authorities. It could as well accept a token and an authorities converter which would be preferable as there is a match between a token and Collection<GrantedAuthority> only for a conversion algorithm.

Reply:
The test support you're adding isn't targeted at integration tests--we aren't trying to exercise or simulate the framework's plumbing, but simply mock its outcomes. You were right to dismiss my idea early-on that the tester provides a raw token value and the test support materializes the resulting `Jwt`. Test support that takes a `Jwt` and converts claims into granted authorities is more of an integration test.

> accept a token and an authorities converter ... as there is a match between token and Collection only for a conversion algorithm.

This is largely the case for any authentication mechanism--the application takes credentials and derives authorities from them. Telling `mockJwt` to use XYZ `GrantedAuthoritiesConverter` would be no different than telling `mockUser` to use ABC `UserDetailsService`.

But this doesn't really matter in the world of tests. In a test, I'm more likely to say "my method requires X, so I will supply Y different varieties of it". And unless I have methods under test that rely on an agreement between my JWT's representation of authorities and Spring Security's representation, any extra support for that at the testing level would be meaningless.

Reply:
> You were right to dismiss my idea early-on

All of my understanding, intentions and implementation have evolved a lot since you started this review, thank you. Regarding `Jwt`, the reason I am opposed to providing public API for it here differ from yours: its construction can't be easily inlined

Could you please detail added value of what you are heading too compared to `.authentication(testingAuthenticationToken)`? With just a little tooling (handy builders for `TestingAuthenticationToken`), it fits perfectly.
I just added such `TestingAuthenticationToken` tooling in [this repo](https://github.com/ch4mpy/spring-addons) with a [sample usage there](https://github.com/ch4mpy/spring-addons/blob/master/spring-security-test-oauth2-addons-showcase/src/test/java/com/c4soft/springaddons/showcase/ShowcaseControllerTests.java) (repo is very small, should be cloned / compiled in an instant)

What I try to provide is tooling for cases requiring `@AuthenticationPrincipal` to return a credible `Jwt`

> The test support you're adding isn't targeted at integration tests

Integration tests covers the decoding and validation of the String payload which is skipped here.

For the unit-test to cover security too, security-configuration is loaded using something like `@WebMvcTest(MyController.class)` or `@ContextConfiguration(classes = MessageServiceTests.SecurityConfiguration.class)`.

IMO, defining the authorities converter clearly is a step in OAuth2 configuration (ideally auto-magically defaulted by the framework and easy to override)

> GrantedAuthoritiesConverter would be no different than telling mockUser to use ABC UserDetailsService.

It would, actually. A user info only holds ID info, not authorizations info while an OAuth2 token holds both (more details on the [RFC-6749](https://tools.ietf.org/html/rfc6749) and https://github.com/spring-projects/spring-security/issues/6807 discussions).

As a consequence, `Collection<GrantedAuthority>` from:
  * a user ID requires calling an external service
  * a Jwt (which is an OAuth2 token, right ?) should be just a matter of reading it

Reply:
> Could you please detail added value of what you are heading to

I added a [PR to your repo](https://github.com/ch4mpy/spring-addons/pull/1) that hopefully outlines what I'm thinking. I added one more controller endpoint that had a slightly more sophisticated expression as well, just for demonstration of the API.

> It would, actually.

In both cases, I'd be invoking test support identical to framework support for translating a cred into a set of authorities. The point of the test is not to confirm that my conversion from cred to authorities works, but that my controller or service works. Neither my controller nor service cares about these details, and I don't want my controller test to fail because my converter has a bug. Where the serialized representation ultimately sits - embedded in the cred or stored in a DB - is immaterial to the point I'm making.

> Expressions can be about anything and actually, 

Agreed. And if a controller needs to query the `scope` claim as well as the granted authorities in the same request, then the tester can and should provide both. If this turns out to be a very common scenario, let's come back and add support to simplify that use case so the tester no longer needs to provide both.

Reply:
> I added a PR to your repo that hopefully outlines what I'm thinking

Thank you for the code, will ease the update of my PR to your repo.

> I don't want my controller test to fail because my converter has a bug

I don't want my test to fail because my Jwt has a bug.
If security config is invalid, test fails. `HttpSecurity` configuration is a step but, again, IMO the choice of the authorities converter is part of security-config too.

> The point of the test is not to confirm that my conversion from cred to authorities works

My point is to provide with credible Jwt (coherent claims and authorities) without requiring the tester to keep it manually in synch. What a confusion if tests start to fail when replacing an expression with an idempotent one or updating a test case after an evolution and having the tests still failing because case value was not updated at all places...

Had understood what you mean and your motivations and that it doesn't match mine. 
I believe we can sum it up as follow:
* you consider GrantedAuthority to be the center of spring-security world when I consider it as a star in the system of OAuth2 Spring implementation
* you consider additional fuzz for the OAuth2 tester to keep in mind how claims are turned into authorities int his app while he writes test cases when I believe it is natural for an OAuth2 dev to think of OAuth2 authorizations (intentionally use the RFC vocabulary here) as token claims (or attributes or parameter or whatever the token type vocabulary uses)
* you are used to `.authorities(new SimpleGrantedAuthority("SCOPE_ROLE_USER"), new SimpleGrantedAuthority("SCOPE_ROLE_ADMIN")` when I prefer `.attribute("scp", "ROLE_USER ROLE_ADMIN")` 

Hope you will understand I'll update my PR on spring security but keep this lib as it is behaving.

Reply:
I think that's a good summary.

I'd only slightly modify the last point:

> you are used to `.authorities(new SimpleGrantedAuthority("SCOPE_ROLE_USER"), new SimpleGrantedAuthority("SCOPE_ROLE_ADMIN")` when I prefer `.attribute("scp", "ROLE_USER ROLE_ADMIN")`

Both of these are easy with either proposal, so I don't think this difference really matters. For me, the tester should simply use whatever the method under test is expecting.

I believe the main difference of opinion is what a tester should have to consider when adding security to the test.

> Hope you will understand I'll update my PR on spring security but keep this lib as it is behaving.

Certainly - I only opened the PR as it seemed a clear way to draw the distinction between our viewpoints. Feel free to leave your library as it is and close the PR when you are through with it.

Reply:
Thanks for all your research, @ch4mpy.

You've convinced me that scopes -> authorities is a reasonable default. As you said with your earlier comment, it's reasonable that a tester would just want to specify scopes and anticipate that the test support would interpret the scopes in what is already a pretty universal way.

I'd like to propose some modifications to your proposal via a few different use cases:

Authenticated with a JWT:
```java
this.mvc.perform(get("/").with(jwt()));
```
This `Jwt` would have a default subject, algorithm, and scope claim. There would be a granted authority that coheres with the scope claim.

Authenticated with a JWT that has scopes:
```java
jwt(jwt -> jwt.claim("scope", "read write"))
```
This `Jwt` would have a default subject and algorithm, and the indicated scope claim. There would be two granted authorities, `"SCOPE_read"` and `"SCOPE_write"`.

The most opinionated is still to provide the authorities, so the following would override any claims in the `Jwt`:
```java
jwt(jwt -> jwt.claim("scope", "not reflected as authorities"))
    .authorities("SCOPE_ADMIN")
```
This `Jwt` would have a default subject and algorithm, and the indicated scope claim. There would be one granted authority, `"SCOPE_ADMIN"`.

But, a conversion strategy still makes sense, so the authorities could be overridden like so:
```java
jwt(jwt -> jwt.claim("scope", scopes))
    .authorities(authoritiesConverter)
```
This `Jwt` would have the default subject and algorithm and the indicated scope claim. There would be whatever granted authorities came out of the conversion.

The point of the consumer is to separate the construction of the `Jwt` from the rest of the API. Separating the concern this way seems more natural than something like `jwt().claim().authorities()` where these concepts are melded together (something that harks back to your comment about wanting to remove the `authorities` method).

It also introduces `JwtBuilder`, which seems very reusable. 

It lines up nicely with other places where this pattern is employed like how http headers and attributes are specified in `WebClient`:

```
WebClient.builder()
    .headers(headers -> headers.setBasicAuth(...))
    .attributes(attributes -> ...)
```

It also lines up with a feature under consideration via #5557 for the `HttpSecurity` DSL.

Reply:
I'll try to put all this together this weekend. Injecting the converter the way you ask is straight forward. Regarding the authorities override,  I still need to figure out a clean way to do it, respecting the "last cal wins" principle. Will do my best for this to be pushed by Monday morning.

Reply:
What I did:
* create a `Jwt.Builder` to set token value, claims and headers
* create a `JwtAuthenticationToken.Builder` to configure the Jwt (`.token(Consumer<Jwt.Builder> jwtBuilderConsumer)`), authentication name and scopes (as specified by [RFC-6749](https://tools.ietf.org/html/rfc6749#section-5.1) : a "scope" claim, value being a spaced separated string)
* modify `JwtAuthenticationTokenTestingBuilder` to extend `JwtAuthenticationToken.Builder`, adding testing default values and authorities extensions
* `.authorities(...)` adds to converted authorities (does not replace it)

What I haven't done: modify `JwtRequestPostProcessor` extend `SecurityContextRequestPostProcessorSupport`. I created https://github.com/spring-projects/spring-security/issues/6857 instead to modify exiting request post processors.

Reason for that is I believe authentication related `MockMvc` post-processors and `WebTestClient` configurers are easier to use if it is (extend) authentication builders (flow API) rather than collaborate with it through `Consumer` interface (`JwtAuthenticationToken.Builder` already collaborating with `Jwt.Builder` through this interface, that would be a lot of indirections.

Sample:
``` java
jwt(jwt -> jwt.claim("foo", "bar")).name("ch4mpy").scopes("TEST"))
```
is easier to read than
``` java
jwt(jwtAuthenticationToken -> jwtAuthenticationToken.token(jwt-> jwt.claim("foo", "bar")).name("ch4mpy").scopes("TEST"))
```

Reply:
Very much agreed that your first example is easier to read than your second.

I'd like to make sure we're on the same page with one thing that your example appears to imply.

I believe we should separate the concerns of configuring a `Jwt` from configuring the rest. So that means once you do:

```java
jwt(jwt -> jwt.claim(...)...)
```

Then, you've completed your `Jwt` configuration. That configuration can propagate forward into other configurations (claims -> authorities), but not the other way around (authorities -> claims).

As such, `name` doesn't make sense unless it's affecting the `JwtAuthenticationToken#getName` method. And `scopes` doesn't make sense, except as a quick way to create granted authorities.

Thus, I think that we should wait on `name` and `scopes` methods for now. It's still quite easy without them:

```java
jwt(jwt -> jwt.subject("ch4mpy"))
    .authorities("SCOPE_TEST")
```

or 

```java
jwt(jwt -> jwt.subject("ch4mpy").claim("scope", "TEST"))
```

While `jwt().name("ch4mpy").scopes("TEST")` is nice and quick, providing a `Jwt.Builder` and `authorities` methods is an MVP that we can iterate on in future PRs.

Reply:
1. I don't see any good reason to limit the token configuration to `jwt(jwtBuilderConsumer)`. It would prevent users from creating method for "common JWT authentication builders" they can further modify in isolated tests (at least token would be immutable). I'm fine to provide with a shortcut to configure the token from `jwt(..)` and `mockJwt(...)` but much reserved to remove the `token(...)` method I wrote on `JwtAuthenticationToken.Builder`
2. this `jwt()` method name is actually aiming at configuring a `JwtAuthenticationToken` which is a `org.springframework.security.core.Authentication` and as so, also a `java.security.Principal`. IMO, this is enough of a reason to expose a method to configure `Principal` "name", specially if you want to expose methods to configure `Authentication` "authorities".
3. `scopes(...)` is just a helper to fill the gap between the spec and Spring implementation:
  * "scope" is specified by https://tools.ietf.org/html/rfc6749#section-5.1 as token meta-data to authorization response
  * Spring `Jwt` is OAuth2 but we can't access token meta-data. Adding a "scope" claim is the usual work-around.